[Pytorch Edge] Black Box Compatibility API

JacobSzwejbka opened the pull request:
Summary:
It would be nice if the compatibility api was just kinda plug and play with no care about the internals of the api at all. Thats what this diff aims to provide.

The general usage would be something like

  RuntimeCompatibilityInfo runtime_info = get_runtime_compatibility_info();
  ModelCompatibilityInfo model_info = get_model_compatibility_info(<model_path>);
  bool compatible = is_compatible(runtime_info, model_info);

Currently RuntimeCompatibilityInfo and ModelCompatibilityInfo are exactly the same, but it seemed feasible to me that they may end up diverging as more information is added to the api (such as a min supported bytecode version being exposed from the runtime).

Test Plan: unit test and ci

Differential Revision: D29624080

iseeyuan commented on test/cpp/jit/test_lite_interpreter.cpp:
```diff
@@ -756,6 +756,36 @@ TEST(LiteInterpreterTest, GetRuntimeOpsAndInfo) {
+  model_ops["aten::add.Scalar"] = OperatorInfo{2};
+  // arbitrarily pick bytecode version 5. This may need to be updated some time
+  // in the future
+  auto model_info = ModelCompatibilityInfo{0x5, model_ops};
```
Consider use `caffe2::serialize::kProducedBytecodeVersion`. Then it's expected to be always compatible.

iseeyuan commented on test/cpp/jit/test_lite_interpreter.cpp:
```diff
@@ -756,6 +756,36 @@ TEST(LiteInterpreterTest, GetRuntimeOpsAndInfo) {
+  AT_ASSERT(!is_compatible(runtime_info, model_info));
+
+  // test trivial failure due to bytecode
+  model_info.bytecode_version = 0x7;
```
Similar above. If bytecode version is bumped to 7 this test would fail. Use `caffe2::serialize::kMaxSupportedBytecodeVersion + 1` instead.

iseeyuan reviewed (approved):
LGTM! Minor suggestions on using version numbers in unit tests.

raziel reviewed (approved):
Thanks!

JacobSzwejbka committed:
[Pytorch Edge] get_model_bytecode int -> uint (#62201)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/62201

change int to uint to be the same type as the runtimes bytecode. Only affects c++ since python doesn't have uints iirc. Also changed the behavior of the functions from returning -1 and a warning to just throw an exception. Wasnt sure what the proper behavior here would be (returning UINT_MAX seemed gross) so feedback is appreciated.

Test Plan: ci

Differential Revision: D29914072

fbshipit-source-id: cce9d47d622f4ce97aaa0082b1cd756ae58a3597

JacobSzwejbka committed:
[Pytorch Edge] Black Box Compatibility API (#61477)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/61477

It would be nice if the compatibility api was just kinda plug and play with no care about the internals of the api at all. Thats what this diff aims to provide.

The general usage would be something like
  < On the Client >
  RuntimeCompatibilityInfo runtime_info = get_runtime_compatibility_info();

  .
  .
  .
  < On the Server >
  ModelCompatibilityInfo model_info = get_model_compatibility_info(<model_path>);
  bool compatible = is_compatible(runtime_info, model_info);

Currently RuntimeCompatibilityInfo and ModelCompatibilityInfo are exactly the same, but it seemed feasible to me that they may end up diverging as more information is added to the api (such as a min supported bytecode version being exposed from the runtime).

Test Plan: unit test and ci

Reviewed By: dhruvbird, raziel

Differential Revision: D29624080

fbshipit-source-id: 6047f4e24d8fbc868cf8109839904978b2d8b622
